### DISCUSSION TITLE
Should multidimensional arrays allow indexes outside of the first 256 cells (8-bit interpreter)

### DISCUSSION THREAD

**Discussion by NeeEoo** [elikaski/BF-it#43] [Q&A]
I'm unsure if this is a bug.

Running this
```c
char aaa[17][17];
aaa[0][0] = 1;
aaa[0][1] = 2;
aaa[0][2] = 3;
aaa[0][3] = 4;
aaa[1][0] = 5;
for(int y = 0; y - 17; y++) {
    for(int x = 0; x - 17; x++) {
        if(aaa[y][x]) {
            printchar(aaa[y][x]+48);
        } else {
            printchar('_');
        }
    }
    printchar('\n');
}
```

Results in this
```
1234_____________
5________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_________________
_1234____________
_5_______________
```

Which is incorrect. It becomes correct on a 16-bit interpreter

**Comment by elikaski** [comment]
I don't think it's a bug, just a wrap around
I think the same thing happens in C when you use a byte as index

**Reply by NeeEoo** [reply]
Could maybe add a flag or something that makes the compiler change the code so it works. By moving x amount of cells in the `get_move_right_index_cells_code`, instead of moving 1 cell per iteration.

It doesn't happen in c. https://godbolt.org/z/rbW5rz

But using `*(*aaa + (char)(y*17 + x))` makes it happen

**Reply by elikaski** [reply]
OK, I understand now
It's because our cell size is 8 bit, like you suggested
If C's registers were 8 bits then it would happen there too I guess 😄 

I think that if we have 8 bits register then it is OK to expect the index to wrap-around
Having said that, I would be happy to have the solution you suggested, which will be enabled with a compilation flag :)

Can you please elaborate on how would that work for index token that is not NUM?

**Reply by NeeEoo** [reply]
It would still work with ID tokens. Since it only needs the dimensions of the array.

```c
int arr[3][6][5];
// ...
int x = 2;
int y = 4;
printchar(arr[2][y][x]);
```

Note: The codes below are a simplification of the compiled code

Old:
```py
index = (2*(6*5))+(y*5)+x

for _ in range(index)
     ptr += 1

printchar(data[ptr])
```

New:
```py
for _ in range(2)
     ptr += 30  # Since dimensions are always known in arrays, we can calculate the product during compile time

for _ in range(y)
     ptr += 5

for _ in range(x)
     ptr += 1

printchar(data[ptr])
```

**Reply by NeeEoo** [reply]
A better thing would be that when #26 is implemented in the compiler (8-bit). I could make it so that it would have a tag named `is_large_array` which makes the array code return a 16 bit index variant.

**Reply by elikaski** [reply]
Good idea
